pkg/controlplane/apiserver/apis.go

```diff
@@ -83,8 +90,55 @@ func (c *CompletedConfig) GenericStorageProviders(discovery discovery.DiscoveryI
 	}, nil
 }
 
+// StorageReadinessHook implements PostStartHook functionality for checking readiness
+// of underlying storage for registered resources.
+//
+// FIXME: Figure out where to put it.
+type StorageReadinessHook struct {
+	checks  map[string]func() error
+	timeout time.Duration
+}
+
+// NewStorageReadinessHook created new StorageReadinessHook.
+func NewStorageReadinessHook(timeout time.Duration) *StorageReadinessHook {
+	return &StorageReadinessHook{
+		checks:  make(map[string]func() error),
+		timeout: timeout,
+	}
+}
+
+func (h *StorageReadinessHook) RegisterStorage(gvr metav1.GroupVersionResource, storage rest.StorageWithReadiness) {
+	// FIXME: handle duplicates
+	h.checks[gvr.String()] = storage.ReadinessCheck
+}
+
+func (h *StorageReadinessHook) check() bool {
+	failedChecks := []string{}
+	for gvr, check := range h.checks {
+		if err := check(); err != nil {
+			failedChecks = append(failedChecks, gvr)
+		}
+	}
+	if len(failedChecks) == 0 {
+		return true
+	}
+	klog.Infof("Storage is not ready for: %v", failedChecks)
+	return false
+}
+
+func (h *StorageReadinessHook) Hook(ctx genericapiserver.PostStartHookContext) error {
+	// FIXME: Handle error propely - timeout should not result in kube-apiserver initialization failure.
```

Comment:
Maybe use PollUntilContextCancel and check for deadline exceeded? E.g. https://goplay.space/#NrdP8_4PnHg?